In [1]:
import numpy as np
import math
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.axes3d import Axes3D
import pandas as pd
import os
%matplotlib inline

Problem: We are to show the feasability of Moore's Law: Moore's Law is the observation that the number 
of transistors on a chip doubles every two years. So we will only be using the year as a predictor.
We will be using the number of transistors as our target, so this will only involve simple linear regression.

In [2]:
#Load data
#btw, I manually added column headers and added the file to the submission
data = pd.read_csv("moore2.csv")

In [3]:
#
#Clean data


data = data.replace(regex=[" ", ",", "~","cca",r"\[...",r"\[.."],value="")
data.year=data.year.astype(int)
data.N=data.N.astype(int)

#Set up Equation
N=len(data)



X= np.zeros((N,1))
X[:,0] = data.year.values

X = np.column_stack((np.ones((N,1)),X)) #102 x 2 Matrix

y = np.zeros((N,1))
y[:,0] = data.N.values #Target Vector

y = np.log2(y) #Tried a couple of differet transforms, but I landed on the log base 2 function


w = np.linalg.solve(X.T.dot(X), X.T.dot(y)) #Gets our model in a 2 x 1 vector
y_hat = X.dot(w)


R2 = 1 - np.sum((y-y_hat)**2)/np.sum((y-y.mean())**2) #Gets our R^2
print(w)
print(R2)

[[-9.88246513e+02]
 [ 5.06448822e-01]]
0.9529442852285763


Our regression after our transform turns out to be
#Number of Transistors = -988 + 0.506 * years
#and our model describes 95.29% of the variance according to our R-squared value.
Because of the transform, log(base2)(X) + 1 = log(base2)(2*X)
So, since the slope is about .5, the amount of transistors is doubling every 2 years
Seemingly, Moore's Law is proven to be correct